torch/testing/_internal/common_utils.py

```diff
@@ -69,6 +69,116 @@
+    there is no need to parametrize over device type, as that is already handled separately.
+    """
+    def _parametrize_test(self, test, generic_cls, device_cls):
+        """
+        Parametrizes the given test function across whatever dimension is specified by the derived class.
+        Tests can be parametrized over any arbitrary dimension or combination of dimensions, such as all
+        ops, all modules, or all ops + their associated dtypes.
+
+        Args:
+            test (fn): Test function to parametrize over
+            generic_cls (class): Generic test class object containing tests (e.g. TestFoo)
+            device_cls (class): Device-specialized test class object (e.g. TestFooCPU); set to None
+                if the tests are not part of a device-specific set
+
+        Returns:
+            Generator object returning 3-tuples of:
+                test (fn): Parametrized test function; must support a device arg and args for any params
+                test_name (str): Parametrized name of the test (e.g. test_bar_opname_int64)
+                param_kwargs (dict): Param kwargs to pass to the test (e.g. {'op': 'add', 'dtype': torch.int64})
+        """
+        raise NotImplementedError
+
+    def __call__(self, fn):
+        fn.parametrize_fn = self._parametrize_test
+        return fn
+
+
+def instantiate_parametrized_tests(generic_cls):
+    """
+    Instantiates tests that have been decorated with a parametrize_fn. This is generally performed by a
+    decorator subclass of _TestParametrizer. The generic test will be replaced on the test class by
+    parametrized tests with specialized names.
+
+    Args:
+        generic_cls (class): Generic test class object containing tests (e.g. TestFoo)
+    """
+    for attr_name in tuple(dir(generic_cls)):
+        class_attr = getattr(generic_cls, attr_name)
+        if not hasattr(class_attr, 'parametrize_fn'):
+            continue
+
+        # Remove the generic test from the test class.
+        delattr(generic_cls, attr_name)
+
+        # Add parametrized tests to the test class.
+        def instantiate_test_helper(cls, name, test, param_kwargs):
+            @wraps(test)
+            def instantiated_test(self, param_kwargs=param_kwargs):
+                test(self, **param_kwargs)
+
+            assert not hasattr(generic_cls, name), "Redefinition of test {0}".format(name)
+            setattr(generic_cls, name, instantiated_test)
+
+        for (test, test_name, param_kwargs) in class_attr.parametrize_fn(
+                class_attr, generic_cls=generic_cls, device_cls=None):
+            instantiate_test_helper(cls=generic_cls, name=test_name, test=test, param_kwargs=param_kwargs)
+
+
+class parametrize(_TestParametrizer):
+    """
+    Decorator for applying generic test parametrizations.
+
+    Args:
+        arg_str (str): String of arg names separate by commas (e.g. "x,y")
```

Comment:
Does pytest allow for naming of the variants?

Reply:
agree that variant naming is nice to have (and I'd consider adding it even if pytest doesn't have it)

Reply:
So my understanding is that pytest's parametrize only works with pytest tests and not unittest (see [here](https://stackoverflow.com/questions/18182251/does-pytest-parametrized-test-work-with-unittest-class-based-tests)).

What this means is that pytest has an internal way of doing variant naming. I don't believe it adds e.g. a `test_func_variant_a()` function to the test class for every variant, in contrast to unittest's modus operandi. So I think we'll have to add some custom mechanism for doing the naming to stay compatible with unittest.

Reply:
If people name the variants I think it's straightforward to generate a test for each of them. If people don't name the variants but the variants are a sequence with a fixed length (like a list or tuple) then we can probably name them _0 through _n. Where naming gets tricky is if we have a variable-length variant generator (like sample inputs), because our test template instantiation happens at "compile time" and we probably don't want to enumerate arbitrary generators just to discover how many elements they have. If people are really keen on using generators maybe we can have them pass the length as a compile-time constant?

Reply:
Added some updates that allow for variant naming. There are two mechanisms for doing so, and whichever is most convenient can be used:
* Specify `name_mappings` (open to other names), which is a list of either dicts of callables that map a parameter value to its string representation. The test suffix is formed by joining parameter string representations with underscores.
* Specify `suffix_fn`, which is a callable that takes in the parameter values and returns a test suffix. This can be used when full control over the suffix is needed and it shouldn't just be an underscore-separated string.

If neither are specified, the default test suffix consists of the lowercase string representation for each parameter value separated by underscores.

Check out the updated PR description for examples!